### ISSUE TITLE
Loading model parameters produces ValueError: variable shape (16,) and value shape (1, 1, 1, 1, 16, 1) are incompatible

### ISSUE THREAD

**Issue by j-c-carr** [j-c-carr/lightcones#2]
From the [log](https://github.com/j-c-carr/lightcones/blob/main/reproducing-results/out/log/2021-05-04_08:35:26.txt), this produces
```
Traceback (most recent call last):
  File "bin/isensee_train.py", line 98, in <module>
    model.load_weights(model_loc)
  File "/home/jcc/Documents/cosmology/lightcones/.env/lib/python3.8/site-packages/tensorflow/python/keras/engine/training.py", line 2234, in load_weights
    hdf5_format.load_weights_from_hdf5_group(f, self.layers)
  File "/home/jcc/Documents/cosmology/lightcones/.env/lib/python3.8/site-packages/tensorflow/python/keras/saving/hdf5_format.py", line 710, in load_weights_from_hdf5_group
    K.batch_set_value(weight_value_tuples)
  File "/home/jcc/Documents/cosmology/lightcones/.env/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py", line 201, in wrapper
    return target(*args, **kwargs)
  File "/home/jcc/Documents/cosmology/lightcones/.env/lib/python3.8/site-packages/tensorflow/python/keras/backend.py", line 3706, in batch_set_value
    x.assign(np.asarray(value, dtype=dtype(x)))
  File "/home/jcc/Documents/cosmology/lightcones/.env/lib/python3.8/site-packages/tensorflow/python/ops/resource_variable_ops.py", line 888, in assign
    raise ValueError(
ValueError: Cannot assign to variable instance_normalization/gamma:0 due to variable shape (16,) and value shape (1, 1, 1, 1, 16, 1) are incompatible
```

**Comment by j-c-carr** [comment]
At first I thought this was due to difference in tensorflow versions, I'm running `tf version 2.4.1` right now. But so is Sam (from the `old_requirements.txt` file. I will try and load a different set of weights and also look around in the source code for tensorflow. It looks like the parameters just need to be resized from the cursed rank-1 array shape

**Comment by j-c-carr** [comment]
Error message from `pip install tensorflow==2.1.0` provides a list of alternative versions >= 2.2.0 I will try 2.2.0 but I'm pretty sure that this is what is causing the problem when loading the weights.

**Comment by j-c-carr** [comment]
Did a clean install of all pip packages using `tf version 2.2.0`, but got some build errors with the `mpi4py` package. This might be what is causing the bug since apparently this is the package that deals with pickling the wieghts

**Comment by j-c-carr** [comment]
Downloaded `python 3.7.6` and `pip 19.3.1` to match the original setup. I was able to install `tensorflow 2.1.0` but I am getting some conflicts in the dependencies between `tensorflow`, `tensorflow-cpu` and `tensorflow-gpu`.

This are leading to more errors in how the model parameters are loaded

**Comment by j-c-carr** [comment]
Missing mpi.h header file which is not letting me download mpi4py package. From [this post](https://stackoverflow.com/questions/26920083/fatal-error-mpi-h-no-such-file-or-directory-include-mpi-h) I was able to download the header file. This seems to be a problem with mpi4py.

Even after downloading the header file I get 
`
/usr/lib/x86_64-linux-gnu/openmpi/include/mpi.h:322:57: error: expected expression before ‘_Static_assert’
      322 | #define THIS_SYMBOL_WAS_REMOVED_IN_MPI30(func, newfunc) _Static_assert(0, #func " was removed in MPI-3.0.  Use " #newfunc " instead.")
`
This doesn't seem to be anything python version / tensorflow related, more so with how `mpi4pi` interacts with my OS

**Comment by j-c-carr** [comment]
Instead of trying to install all requirements, I just tried to install the basic ones (ie. I avoided installing `tensorflow-gpu` and `tensorflow-cpu`, because they had dependency conflicts with the `tensorflow` package. These are saved in `in/base_requirements.txt`

I was able to install a slightly older version of mpi4py. `conda 4.8.3` is still unavailable (I can't find that version of conda on their distributions page). My version of `conda` is 4.10.0, which the closest to 4.8.3 that I can find it so I stuck with this version.

When I run `bin/isensee_train.py`, I get the same error (the same as in the title of this issue) when I try to load the model weights. I think there could be two types of problems:
1. There is a mismatch between mpi4py version `3.0.3` (which I have), and version `3.0.1` (which the model was saved with). This could be what is causing the model-weights file to not be read in correctly.
2. There are other packages among `in/requirements_orig.txt` that need to be installed (for example, the tensorflow-cpu package)

I'm leaning towards problem (1) because I am able to reproduce the exact same result as before, when I installed all of the packages in `in/requirements_orig.txt` using python 3.8. 

Also, I'm wondering if it is even worth it to be able to load the model wieghts, if the next step of the project is to try and train the net and generate new parameters anyways.

Lastly, are the differences between `in/base_requirements.txt` and `in/requirements_orig.txt`.
[2020-05-04-1423_diff.txt](https://github.com/j-c-carr/lightcones/files/6423193/2020-05-04-1423_diff.txt)

**Comment by j-c-carr** [comment]
It seems like this is a common issue with tensorflow as documented [her](https://stackoverflow.com/questions/63810835/tensorflow-2-x-cannot-load-trained-model-in-h5-format-when-using-embedding-colu), [here](https://stackoverflow.com/questions/61464888/tensorflow-error-valueerrorshapes-s-and-s-are-incompatible-self-other), [here](https://datascience.stackexchange.com/questions/49546/keras-load-pre-trained-weights-shape-mismatch), and [here](https://stackoverflow.com/questions/56169945/incompatible-shapes-in-keras-when-loading-a-model-with-custom-layer). Most solutions involve either upgrading tensorflow or supplying different parameters when training the model.

**Comment by j-c-carr** [comment]
I ended up patching the source code, at `../.env/lib/python3.7/site-packages/tensorflow_core/python/ops/resource_variable_ops.py`,
line 810, 
```
      try:
          self._shape.assert_is_compatible_with(value_tensor.shape)
      except ValueError:
          value_tensor = tf.squeeze(value_tensor)
          self._shape.assert_is_compatible_with(value_tensor.shape)
```
This essentially removes the axes of size 1. However, I had to import the `tensorflow` package inside of the `tensorflow.core` source code, so I'm not sure if this will cause unexpected errors down the line. It does fix this particular issue so I'll close it for now.
